wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs

```diff
@@ -0,0 +1,138 @@
+                     wallets)
+import qualified Cardano.Wallet.Kernel.Keystore as Keystore
+import           Cardano.Wallet.Kernel.PrefilterTx (prefilterUtxo)
+import           Cardano.Wallet.Kernel.Types (WalletId (..))
+import           Cardano.Wallet.Kernel.Util (getCurrentTimestamp)
+
+import           Test.QuickCheck (Arbitrary (..), oneof)
+
+
+data CreateWalletError =
+      CreateWalletFailed HD.CreateHdRootError
+      -- ^ When trying to create the 'Wallet', the DB operation failed.
+
+instance Arbitrary CreateWalletError where
+    arbitrary = oneof []
+
+instance Buildable CreateWalletError where
+    build (CreateWalletFailed dbOperation) =
+        bprint ("CreateWalletUnknownHdAccount " % F.build) dbOperation
+
+instance Show CreateWalletError where
+    show = formatToString build
+
+instance Exception CreateWalletError
+
+{-------------------------------------------------------------------------------
+  Wallet Creation
+-------------------------------------------------------------------------------}
+
+-- | Creates a new HD 'Wallet'.
+createHdWallet :: PassiveWallet
+             -> Mnemonic nat
+             -- ^ The set of words (i.e the mnemonic) to generate the initial seed.
+             -- See <https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#From_mnemonic_to_seed>
+             -- This Kernel function is agnostic in the number of words, and it's
+             -- wallet layer's responsibility to make sure that invalid sizes are
+             -- rejected.
+             -> PassPhrase
+             -- ^ The spending password to encrypt the 'SecretKey' for the
+             -- newly-generated wallet. If the user didn't specify any, the
+             -- empty 'PassPhrase' is used.
+             -> AssuranceLevel
+             -- ^ The 'AssuranceLevel' for this wallet, namely after how many
+             -- blocks each transaction is considered 'adopted'. This translates
+             -- in the frontend with a different threshold for the confirmation
+             -- range (@low@, @medium@, @high@).
+             -> WalletName
+             -- ^ The name for this wallet.
+             -> IO (Either CreateWalletError HdRoot)
+createHdWallet pw mnemonic spendingPassword assuranceLevel walletName = do
+    -- STEP 1: Generate the 'EncryptedSecretKey' outside any acid-state
+    -- transaction, to not leak it into acid-state's transaction logs.
+    let (_, esk) = safeDeterministicKeyGen (BIP39.mnemonicToSeed mnemonic) spendingPassword
+    -- STEP 2: Atomically generate the wallet and the initial internal structure in
+    -- an acid-state transaction.
+    let newRootId = eskToHdRootId esk
+    res <- createWalletHdRnd pw
+                             (spendingPassword /= emptyPassphrase)
+                             walletName
+                             assuranceLevel
+                             esk
+                             mempty
```

Comment:
Add a comment here and in the type signature specifying the invariants of this function, i.e. that the `Mnemonic` shouldn't have been used anywhere else before, otherwise this won't work as it passes in the empty Utxo and thus `prefilterUtxo` won't work correctly.